msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/DiscussionForumTool.java

```diff
@@ -2224,6 +2236,220 @@ public String processActionToggleDisplayExtendedDescription()
+          } catch (IdUnusedException e) { //this might happen if the event was removed from Calendar without this Forum data being changed, so Forums would have a record for a Calendar event that doesn't exist. We can just create it.
+            end = targetCalendar.addEvent();
+          }
+          targetCalendar.removeEvent(end);
+        }
+        if (forumItem instanceof DiscussionTopic) { //put the null begin/end IDs in the forumItem itself as well, depending on type.
+          ((DiscussionTopic) forumItem).setCalendarBeginId(null);
+          ((DiscussionTopic) forumItem).setCalendarEndId(null);
+        } else if (forumItem instanceof DiscussionForum) {
+          ((DiscussionForum) forumItem).setCalendarBeginId(null);
+          ((DiscussionForum) forumItem).setCalendarEndId(null);
+        }
+      }
+    } catch (IdUnusedException e) { //for getCalendar, if the Calendar doesn't exist.
+      e.printStackTrace();
+    } catch (PermissionException e) { //for addEvent, if the user can't edit a calendar.
+      e.printStackTrace();
+    } catch (InUseException e) {  //for getEditEvent, if it's already being edited.
+      e.printStackTrace();
+    }
+  }
+
+  public Boolean getDoesSiteHaveCalendar(){  //if any Calendar data exists for this site, this method returns True. It's not necessarily based on the literal presence of the Calendar tool in the site, but on Calendar-related data.
+    try { //just look for a usable calendar with this site's ID.
+      this.calendarService.getCalendar("/calendar/calendar/" + getContextSiteId().replace("/site/", "") + "/main");
+    } catch (IdUnusedException e) { //this would be the normal Catch case...when the calendar we tried to get just doesn't exist.
+      log.debug(" No Calendar data exist for this site.");
+      return false;
+    } catch (PermissionException e) { //this just means the user can't edit. Probably rare.
+      log.debug(" User does not have permission to edit this calendar.");
+      return false;
+    }
+    return true;  // if no exceptions get thrown, return True...either usable Calendar data exists, or the Calendar tool is in the site.
+  }
+
+  private Collection<Group> getAllowedGroups(Set membershipItems) {  //get the groups allowed to see a Forum item's related Calendar events.
+    if (membershipItems == null) {
+      return new ArrayList<>();
+    }
+    Iterator traverseMembershipItems = membershipItems.iterator();
+    Collection<Group> output = new ArrayList<Group>();
+    ArrayList<SiteGroupBean> availableGroups = (ArrayList<SiteGroupBean>) getSiteGroups();
+    ArrayList<String> availableGroupNames = new ArrayList<String>();
+    for (int count = 0; count < availableGroups.size(); count++){ //build an array of just the group names.
+      availableGroupNames.add(availableGroups.get(count).getGroup().getTitle());
+    }
+    while (traverseMembershipItems.hasNext()) {
```

Comment:
I'd just use for (DBMembershipItemImpl now : traverseMembershipItems) {

Reply:
In conjunction with the previous change you requested, this won't compile. Something about typing the Iterator makes Maven unable to understand the for-each construction here.

Reply:
Took me a while to figure out how to get this to state where it would compile, but I now have it looking much more like the way you've suggested here. By typing this method's parameter to plain DBMembershipItem [rather than Impl], and typing the membershipItems Set that gets sent to it above [in the main processActionSendToCalendar method] the same way, I was able to iterate over the membershipItems Set itself with no Iterator:
`for(DBMembershipItem now: membershipItems)`
This compiled fine and then worked fine too.